internal/frr/frr.go

```diff
@@ -45,67 +62,110 @@ func (f *FRR) ApplyConfig(config *Config) error {
 var debounceTimeout = 3 * time.Second
 var failureTimeout = time.Second * 5
 
-func NewFRR(ctx context.Context, logger log.Logger, logLevel logging.Level) *FRR {
+func NewFRR(ctx context.Context, onConfigChanged ConfigChanged, logger log.Logger, logLevel logging.Level) *FRR {
 	res := &FRR{
-		reloadConfig: make(chan reloadEvent),
-		logLevel:     logLevelToFRR(logLevel),
+		reloadConfig:    make(chan reloadEvent),
+		logLevel:        logLevelToFRR(logLevel),
+		onConfigChanged: onConfigChanged,
 	}
 	reload := func(config *Config) error {
 		return generateAndReloadConfigFile(config, logger)
 	}
 
 	debouncer(ctx, reload, res.reloadConfig, debounceTimeout, failureTimeout, logger)
-	reloadValidator(ctx, logger, res.reloadConfig)
+	res.pollStatus(ctx, logger)
 	return res
 }
 
-func reloadValidator(ctx context.Context, l log.Logger, reload chan<- reloadEvent) {
-	var tickerIntervals = 30 * time.Second
-	var prevReloadTimeStamp string
+func (f *FRR) GetStatus() Status {
+	f.RLock()
+	defer f.RUnlock()
+	return f.Status
+}
 
+func (f *FRR) pollStatus(ctx context.Context, l log.Logger) {
+	var tickerIntervals = 30 * time.Second
 	ticker := time.NewTicker(tickerIntervals)
 	go func() {
-		select {
-		case <-ticker.C:
-			validateReload(l, &prevReloadTimeStamp, reload)
-		case <-ctx.Done():
-			return
+		for {
+			select {
+			case <-ticker.C:
+				status, err := fetchStatus()
+				if err != nil {
+					level.Error(l).Log("op", "fetch status", "error", err)
+					break
+				}
+				if status.updateTime == f.Status.updateTime {
+					break
+				}
+				if status.LastReloadResult != ReloadSuccess {
+					level.Error(l).Log("op", "fetch status", "error", err)
+					f.reloadConfig <- reloadEvent{useOld: true}
+				}
+				f.Lock()
+				f.Status = status
+				f.Unlock()
+				if f.onConfigChanged != nil {
+					f.onConfigChanged()
+				}
+			case <-ctx.Done():
+				return
+			}
 		}
 	}()
 }
 
-const statusFileName = "/etc/frr_reloader/.status"
+const (
+	statusFileName    = "/etc/frr_reloader/.status"
```

Comment:
it's a bit confusing now that this file is called status when what it does is holding a timestamp, can you change it?